LLL: fix handling of "sha3" expression

benjaminion opened the pull request:
When PR #2317 changed the EVM opcode from SHA3 to KECCAK256 it broke the `(sha3 loc len)` expression in LLL. This PR fixes things while allowing existing code using the sha3 expression (such as the ENS registrar) to continue to compile.  I.e. both `(keccak256 loc len)` and `(sha3 loc len)` may be used, and the existing related sha3 macros continue to work.

axic commented:
Can you please add an end to end test for both `keccak256` and `sha3`? That should have caught this in the first place 😉

axic commented on liblll/CompilerState.cpp:
```diff
@@ -55,10 +55,10 @@ void CompilerState::populateStandard()
 	"(def 'create (value code) { [0]:(msize) (create value @0 (lll code @0)) })"
 	"(def 'create (code) { [0]:(msize) (create 0 @0 (lll code @0)) })"
+	"(def 'sha3 (loc len) (keccak256 loc len))"
 	"(def 'sha3 (val) { [0]:val (sha3 0 32) })"
```
I think these two might be in conflict?

benjaminion replied:
No, no conflict. The macro expansion takes into account the number of arguments so the two sha3 macros are distinguished correctly.

benjaminion commented:
Three test cases added:

 - kekkac256
 - sha3 macro with 2 args
 - sha3 macro with 1 arg

axic commented on test/liblll/EndToEndTest.cpp:
```diff
@@ -279,6 +279,43 @@ BOOST_AUTO_TEST_CASE(assembly_codecopy)
+	)";
+	compileAndRun(sourceCode);
+	BOOST_CHECK(callFallback() == encodeArgs(
+		fromHex("b10e2d527612073b26eecdfd717e6a320cf44b4afac2b0732d9fcbe2b7fa0cf6")));
```
@chriseth can you please comment on the coding style you use for these kind of statements?

axic commented on test/liblll/EndToEndTest.cpp:
```diff
@@ -279,6 +279,43 @@ BOOST_AUTO_TEST_CASE(assembly_codecopy)
+		(returnlll
+			(seq
+				(mstore 0x00 0x01)
+				(return (keccak256 0x0 0x20))))
```
Nitpicking: can you use `0x00`?

benjaminion replied:
Fair comment! I've amended the commit accordingly.

axic commented:
@benjaminion can you please rebase this? (Since the other LLL changes were merged this is in conflict.)

benjaminion committed:
LLL: fix handling of "sha3" expression

When PR #2317 changed the EVM opcode from SHA3 to KECCAK256 it broke the `(sha3 loc len)` expression in LLL. This PR fixes things while allowing existing code using the sha3 expression (such as the ENS registrar) to continue to compile.  I.e. both `(keccak256 loc len)` and `(sha3 loc len)` may be used, and the existing related sha3 macros continue to work.

Three end-to-end test cases have been added for kekkac256 and sha3.

benjaminion commented:
@axic 

Rebase done. Should be good - all LLL tests succeed.

axic commented:
@benjaminion would you be interested adding tests for every internal macro in a separate PR? It would be nice catching these errors when we make changes 😉

axic commented:
Also thanks for this, merged!

benjaminion commented:
@axic 

> would you be interested adding tests for every internal macro in a separate PR?

Sure. It'll be a week or two, but happy to take this one.
